enterprise/server/remote_execution/copy_on_write/copy_on_write.go

```diff
@@ -1355,3 +1353,93 @@ func checkBounds(opName string, storeSize int64, p []byte, off int64) error {
 	}
 	return nil
 }
+
+// BoundedStack is a concurrency-safe stack with bounded size: when the
+// capacity of the stack is exceeded, the least recently added item ("bottom")
+// of the stack is removed.
+//
+// Its intended usage is to prioritize the most recently added items for eager
+// chunk fetching.
+//
+// It exposes a channel C which streams a value whenever a value is pushed on
+// the stack. Example:
+//
+//	for range stack.C {
+//		val, ok := stack.Pop()
+//		if !ok {
+//			// A receive from C normally guarantees that a value is ready,
+//			// but it's best to always check `ok` when popping from the stack.
+//			continue
+//		}
+//		// Do something with val...
+//	}
+type BoundedStack[T any] struct {
+	C chan struct{}
+
+	mu     sync.Mutex
+	buffer []T
+	top    int
+	size   int
+}
+
+func NewBoundedStack[T any](capacity int) (*BoundedStack[T], error) {
+	if capacity <= 0 {
+		return nil, status.InvalidArgumentError("bounded stack capacity must be > 0")
+	}
+	return &BoundedStack[T]{
+		C:      make(chan struct{}, 1),
```

Comment:
Is the capacity supposed to be capacity and not 1?

Reply:
It should work as-is, because we replenish the buffer in Pop() whenever it still has items after popping. Setting it to `capacity` lets us remove that replenishing logic though so I removed it. This has a bit of extra memory cost but should be fine for now.